src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs

```diff
@@ -2149,6 +2159,111 @@ static void assertBindIdentifierTargets(InMethodBinder? inMethodBinder, Concurre
+                    continue;
+                }
+                var syntax = token.Parent;
+                switch (syntax)
+                {
+                    case IdentifierNameSyntax identifierName when identifierName.Identifier == token:
+                    case GenericNameSyntax genericName when genericName.Identifier == token:
+                    case TupleElementSyntax tupleElement when tupleElement.Identifier == token:
+                    case FromClauseSyntax fromClause when fromClause.Identifier == token:
+                    case LetClauseSyntax letClause when letClause.Identifier == token:
+                    case JoinClauseSyntax joinClause when joinClause.Identifier == token:
+                    case JoinIntoClauseSyntax joinIntoClause when joinIntoClause.Identifier == token:
+                    case QueryContinuationSyntax queryContinuation when queryContinuation.Identifier == token:
+                    case LocalFunctionStatementSyntax localFunctionStatement when localFunctionStatement.Identifier == token:
+                    case VariableDeclaratorSyntax variableDeclarator when variableDeclarator.Identifier == token:
+                    case SingleVariableDesignationSyntax singleVariable when singleVariable.Identifier == token:
+                    case LabeledStatementSyntax labeledStatement when labeledStatement.Identifier == token:
+                    case ForEachStatementSyntax forEachStatement when forEachStatement.Identifier == token:
+                    case CatchDeclarationSyntax catchDeclaration when catchDeclaration.Identifier == token:
+                    case TypeParameterSyntax typeParameter when typeParameter.Identifier == token:
+                    case ParameterSyntax parameter when parameter.Identifier == token:
+                        switch (token.Text)
+                        {
+                            case "field":
+                                if (method.AssociatedSymbol is PropertySymbol { IsIndexer: false })
+                                {
+                                    // Report "field" conflict with keyword.
+                                    reportConflict(syntax, token, diagnostics);
+                                }
+                                break;
+                            case "value":
+                                if (method.MethodKind == MethodKind.PropertySet)
+                                {
+                                    // Record the potential "value" conflict, and report conflicts later,
+                                    // after dropping any that refer to the implicit parameter.
+                                    valueIdentifiers ??= PooledDictionary<SyntaxNode, SyntaxToken>.GetInstance();
+                                    valueIdentifiers.Add(syntax, token);
+                                }
+                                break;
+                        }
+                        break;
+                    default:
```

Comment:
TBH i find hte approach taken in this pr extremely surprising.  It's not how i would have expected this to be done.

Instead, i would expect us to treat his how we treat query-keywords or the await-keyword.  Specifically, the parser should enter into a new contextual parsing state when we enter into these specific contexts.  

Within that appropriate context, we then have the following behavior:

1. in c# 13, the word becomes a keyword, and the new semantics take effect.
2. prior to c# 12, when eating an identifier, if we have one of these keywords, we convert it to an identifier, and report the diagnostics about the potential future issues. 

I think we also need this for incremental parsing to work properly.  If we reinterpret the statements from an accessor into a different context (or vice-versa), then we need to reparse since `field/value` will change from keywords to identifiers.  

--

TLDR: as a contextual keyword, this should work the same way as async/await and query-keywords.

Reply:
Note: i was able to get us additional bits on NodeFlags.  So we can store this info there as well for incremental parsing purposes.  A

Reply:
> 2. prior to c# 12, when eating an identifier, if we have one of these keywords, we convert it to an identifier, and report the diagnostics about the potential future issues.

The context from the parser might help, but the diagnostic may need to be reported in binding since we only want a diagnostic when the identifier would bind differently as a keyword.

Reply:
my approach supports that.  these would be identifiers whose contextual kind would either be 'nothing' (which means it's a legal usage of `field/value` as an identifier), or it has a contextual kind of SyntaxKind.FieldKeyword or SyntaxKind.ValueKeyword.  IN that case, in binding you report a warning saying this will change meaning.

basically, the parser has the core logic for figuring out if it's an identifier or keyword.  if it's a true identifier, great, you get an identifier and everything works as normal.  If it's contextual keyword, then in C# 13 you get a keyword and the new semantics.  In C# 12 you get an identifier, marked as being a contextual keyword, and the binder warns you.  

Note: i don't actually see why it as to be the binder that needs to care.  Sicne C# 13 will have these rules set up at the syntactic level, it feels fine for the parser to warn here for me.  But i'm fine with it beign the binder if you prefer that.

Reply:
I think what Chuck is getting at is that, for example, there's no problem with using identifier 'value' to reference the setter value parameter. The parsing will technically change from C# 12 to 13, but it will still refer to the same thing in its keyword form.

So in order to know whether to give diagnostics, compiler needs to know whether 'value' is referring to that parameter, or referring to something else such as a 'value' local declared in a nested function.

FWIW, in the case I just described, I feel ok with saying you did something really weird and you don't get to have this warning. It's the .01% case.

Maybe we should only warn for use of identifier 'field' in property accessors and just do it in the parser.

Reply:
my approach would work for this afaict.  the node indicates the information needed by the binder to warn.